### PR TITLE
build: update to clang 1.4.0 and only run clang format on changed files

### PR DESCRIPTION
**Pull request by josephperrott**
Update to clang@1.4.0 to gain support for optional changing and nullish
coalescing.  Because this would trigger a change on >1800 files in the
repository, also changes our format enforcement to only be run against
changed files.  This will allow us to incrementally roll out the value
add of the upgraded clang format.

Additionally, this cuts ~2 minutes off the time it takes for the `lint` job to run as not as many files are being checked for format.

### CODE REVIEW

File: gulpfile.js [hunk 1]
```diff
@@ -26,26 +26,31 @@ function loadTask(fileName, taskName) {
...
 gulp.task('format:untracked', loadTask('format', 'format-untracked'));
+gulp.task('format:untracked:enforce', loadTask('format', 'enforce-untracked'));
 
-// Format only the changed, tracked source code files.
+// Changed, tracked source code files.
 gulp.task('format:diff', loadTask('format', 'format-diff'));
+gulp.task('format:diff:enforce', loadTask('format', 'enforce-diff'));
 
-// Format only changed lines based on the diff from the provided --branch
+// Changed lines based on the diff from the provided --branch
```
Comment (gkalpak) [inline gulpfile.js]: I know the wording pre-existed, but does this really only affect changed lines? I thought it affects whole changed files (i.e. whole files as long as they contain a changed line).

File: gulpfile.js [hunk 2]
```diff
@@ -26,26 +26,31 @@ function loadTask(fileName, taskName) {
...
 // argument (or `master` by default).
 gulp.task('format:changed', ['format:untracked', 'format:diff']);
+gulp.task('format:changed:enforce', ['format:untracked:enforce', 'format:diff:enforce']);
 
 // Alias for `format:changed` that formerly formatted all files.
 gulp.task('format', ['format:changed']);
 
-gulp.task('lint', ['format:enforce', 'validate-commit-messages']);
+gulp.task('lint', ['format:changed:enforce', 'validate-commit-messages']);
```
Comment (gkalpak) [inline gulpfile.js]: I'm afraid this won't work correctly on non-master branches (we need to somehow explicitly pass the correct branch).

**Review by gkalpak — COMMENTED** [review]
Yay! Sooooo looking forward to getting this in 😍

File: gulpfile.js (hunk 2 above)
Comment (josephperrott) [reply gulpfile.js]: We are able to pass the branch as a parameter `--branch 9.0.x`.

We should be able to set this up on CI to happen as well.

File: yarn.lock [hunk 3]
```diff
@@ -3566,19 +3566,19 @@ cjson@^0.3.1:
   dependencies:
     json-parse-helpfulerror "^1.0.3"
 
-clang-format@1.0.41:
-  version "1.0.41"
-  resolved "https://registry.yarnpkg.com/clang-format/-/clang-format-1.0.41.tgz#28552e50b0c9e44d23f85aebe3d2fbd06c93620c"
-  integrity sha1-KFUuULDJ5E0j+Frr49L70GyTYgw=
+clang-format@^1.0.32:
+  version "1.2.3"
```
[the remark this answers is not in the document]
Comment (josephperrott) [reply yarn.lock]: Agreed, this is actually just what yarn did automatically.  I have updated it as requested.

File: gulpfile.js (hunk 2 above)
Comment (gkalpak) [reply gulpfile.js]: Yeah, sorry, that's what I meant :grin:
In order for this to work on CI (both for PR builds and upstream builds), we need to pass the correct branch.

Comment (josephperrott) [reply gulpfile.js]: Done, I set it up to run from the base commit sha on CI.

**Review by gkalpak — APPROVED** [review]
:shipit:

**Pull request closed without merging**